srml/elections-phragmen/src/lib.rs

```diff
@@ -0,0 +1,1259 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Phragmen Election Module.
+//!
+//! An election module based on sequential phragmen.
+//!
+//! ### Term and Round
+//!
+//! The election happens in _rounds_: every `N` blocks, all previous members are retired and a new
+//! set is elected (which may or may not have an intersection with the previous set). Each round
+//! lasts for some number of blocks defined by `TermDuration` storage item. The words _term_ and
+//! _round_ can be used interchangeably in this context.
+//!
+//! `TermDuration` might change during a round. This can shorten or extend the length of the round.
+//! The next election round's block number is never stored but rather always checked on the fly.
+//! Based on the current block height and `TermDuration`, `BlockNumber % TermDuration == 0` will
+//! always trigger a new election round.
+//!
+//! ### Voting
+//!
+//! Voters can vote for as many of the candidates by providing a list of account ids. Invalid votes
+//! (voting for non-candidates) are ignored during election. Yet, a voter _might_ vote for a
+//! future candidate. Voters reserve a bond as they vote. Each vote defines a `value`. This amount
+//! is locked from the account of the voter and indicated the weight of the vote. Voters can update
+//! their votes at any time by calling `vote()` again. This keeps the bond untouched but can
+//! optionally change the locked `value`. After a round, votes are kept and might still be valid
+//! for further rounds. A voter is responsible to call `remove_voter` upon once they are done to
+//! have their bond back and remove the lock.
+//!
+//! ### Candidacy and Members
+//!
+//! Candidates also reserve a bond as they submit candidacy. A candidate cannot take their candidacy
+//! back. A candidate can end up in one of the below situations:
+//!   - **Winner**: A winner is kept as a _member_. They must still have a bond in reserve
+//!		and they are automatically counted as a candidate for the next election.
+//!   - **Runner-up**: Runner ups are the best candidates immediately after the winners. The number
```

Comment:
how do runners-up work with phragmen?

Reply:
yeah, it is a bit quirky and so far I only introduced it for the sake of having runner-ups, as requested during reviews. 

I do this to compute it without a lot of overhead in a way that doesn't seem horribly wrong: 

- if you want `m` members and `r` runnerups, call into phragmen to give you a result set of size `m + r`.
- Sort them based on the backed stake and take the best `m` as winners and the rest as runner ups. 

Though, as mentioned, I'm actually more leaned toward having no runner ups in this election. It would make it substantially simpler and as you pointed out the concept doesn't really exist in Phragmen by default.

Reply:
right; let's just go with the no-runners-up case. not sure what i was thinking during the reviews.